Dimensional Analysis Improvements

mgreminger opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

#### Brief description of what is fixed or changed

In working on my [browser based app](https://github.com/mgreminger/EngineeringPaper) that does symbolic computations with dimensional analysis (powered by SymPy through Pyodide), I've come across several dimensional analysis edge cases that weren't working as expected. I've fixed those on my fork and would like to upstream these fixes.  

Summary of changes:
* The `reduce` call that gets triggered for function args didn't have an initial value of 1 so it failed for dimensionless arguments in cases like cos(1). 
* Numbers with a complex component or constants such as pi and E were not handled, they are now treated as dimensionless by default
* The default for functions previously was to allow any dimensions for the input args and then return dimensionless, the default behavior has been changed to enforce all function args to be dimensionless and to return dimensionless (this is a safer default, it didn't make sense for cos(length), cos(time), cos(angle) to all work and return dimensionless
* The functions sin, cos, tan, cot, sec, and csc are treated as a special case and are allowed to have a dimensionless or angle input argument (there may need to be other special case functions in the future but the above default should work for most cases included logarithmic functions)
* Exponents must now be dimensionless (cases such as length**length were allowed previously)


##### Old Behavior
```python
>>> import sympy
>>> sympy.__version__
'1.7.1'
>>> from sympy.physics.units.definitions.dimension_definitions import length, angle, time
>>> from sympy.physics.units.systems.si import dimsys_SI
>>> dimsys_SI.get_dimensional_dependencies(sympy.cos(1))
TypeError: reduce() of empty sequence with no initial value
>>> dimsys_SI.get_dimensional_dependencies(sympy.pi)
TypeError: Type <class 'sympy.core.numbers.Pi'> not implemented for get_dimensional_dependencies
>>> dimsys_SI.get_dimensional_dependencies(sympy.sqrt(-1))
TypeError: Type <class 'sympy.core.numbers.ImaginaryUnit'> not implemented for get_dimensional_dependencies
>>> dimsys_SI.get_dimensional_dependencies(sympy.cos(length))
{}
>>> dimsys_SI.get_dimensional_dependencies(length**length)
{'length': Dimension(length, L)}
```

##### New Behavior
```python
>>> import sympy
>>> from sympy.physics.units.definitions.dimension_definitions import length, angle, time
>>> from sympy.physics.units.systems.si import dimsys_SI
>>> dimsys_SI.get_dimensional_dependencies(sympy.cos(1))
{}
>>> dimsys_SI.get_dimensional_dependencies(sympy.pi)
{}
>>> dimsys_SI.get_dimensional_dependencies(sympy.sqrt(-1))
{}
>>> dimsys_SI.get_dimensional_dependencies(sympy.cos(length))
TypeError: The input argument for the function cos must be dimensionless or have dimensions of angle.
>>> dimsys_SI.get_dimensional_dependencies(length**length)
TypeError: The exponent for the power operator must be unitless.
```

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:


or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.units
  * Fix bug when input argument to a function is an integer
  * Change default behavior for functions (all function arguments must be dimensionless by default)
  * sin, cos, tan, cot, sec, and csc functions may have a dimensionless or angle input argument
  * Exponents must now be dimensionless
  * Constants such as pi and E are now treated as dimensionless (raised exception previously)
  * Numbers with an imaginary component are now treated as dimensionless (raised exception previously)
<!-- END RELEASE NOTES -->

mgreminger committed:
Fix multiple dimension bugs

Fix cos(1) bug, make symbolic numbers such as pi and E dimensionless, fix dimension handling for trig and inverse trig functions and raise for unsupported functions.

Functions  that still need to be implemented: log and exp functions, and hyperbolic trig functions

mgreminger committed:
Add dimensional analysis for log and exp functions

mgreminger committed:
Make the default for functions to be unitless in and out

This is a safe default. Previously the units of the input args were ignored by default. Remove unecessary imports, make sure the exponent of the power operator is unitless.

mgreminger committed:
Remove debugging print statement

mgreminger committed:
Added tests for dimensional analysis fixes

mgreminger committed:
Remove extra newline at end of file

mgreminger committed:
Fix error when result evaluates to a complex number

mgreminger committed:
Remove trailing whitespace

mgreminger committed:
Add dim analysis tests for 2-param calls to log

mgreminger committed:
Fix error message (dimensionless instead of unitless)

oscarbenjamin commented:
CC @Upabjojr

Upabjojr commented on sympy/physics/units/dimensions.py:
```diff
@@ -437,18 +438,32 @@ def _get_dimensional_dependencies_for_name(self, name):
                 return self.get_dimensional_dependencies(result)
             elif result.func == name.func:
-                return {}
+                if name.func in (sin, cos, tan, cot, sec, csc):
```
this should be `isinstance(name, (sin, cos, tan, cot, sec, csc))`, because using `name.func` is not aware of subclassing and may change behaviour in the future. There are some discussions somewhere on why it's always better to use `isinstance`.

mgreminger replied:
Makes sense, I've updated it to reflect this change.

Upabjojr commented on sympy/physics/units/tests/test_dimensions.py:
```diff
@@ -100,8 +100,7 @@ def test_Dimension_mul_div_exp():
     assert dimsys_SI.get_dimensional_dependencies(velo ** -1.5) == {"length": -1.5, "time": 1.5}
 
-    length_a = length**"a"
-    assert dimsys_SI.get_dimensional_dependencies(length_a) == {"length": Symbol("a")}
```
why is this being removed?

mgreminger replied:
I've changed the behavior to require exponents to be dimensionless so I've updated the test to reflect this. As far as I understand, an exponent should be dimensionless.

Upabjojr replied:
Yes, but it can still be a symbol. The symbol may still me dimensionless.

mgreminger replied:
Okay, I see, thanks for the clarification. I've updated the behavior to allow a Symbol, in addition to a Number, as an exponent. A Dimension as an exponent raises an excpetion. I restored the original test that you highlighted to make sure a Symbol works as an exponent.

mgreminger replied:
Fixed a regression I introduced that didn't allow a dimensionless expression as an exponent. This has been fixed and I added an additional test case for this regression.

mgreminger committed:
Use isinstance for checking trig functions

Upabjojr commented on sympy/physics/units/dimensions.py:
```diff
@@ -437,18 +438,32 @@ def _get_dimensional_dependencies_for_name(self, name):
                 return self.get_dimensional_dependencies(result)
             elif result.func == name.func:
-                return {}
+                if isinstance(name, (sin, cos, tan, cot, sec, csc)):
```
Well, there is an even better idea actually. There is an abstract class `TrigonometricFunction` (in `sympy.functions.elementary.trigonometric`), all these classes are subclasses of `TrigonometricFunction`. Maybe simply better `isinstance(name, TrigonometricFunction)`.

Upabjojr replied:
(sorry for not telling before about that)

mgreminger replied:
Yes, good point, that makes it simpler. I've made the change.

mgreminger committed:
Use TrigonometricFunction class to check for trig func

mgreminger committed:
Allow exponent to be number or Symbol but not Dimension

mgreminger committed:
Removed extra blank line

mgreminger committed:
Fix regression in previous commit

The previous commit created a regression where exponents that were a dimensionless expression were not allowed. This has been fixed and a test case has been added.

Upabjojr commented:
Thanks for your contribution!
